File: fav/nodes/scenario_node.py
#!/usr/bin/env python3
import math
import os
from collections import deque
from dataclasses import dataclass
from statistics import mean

import rclpy
import yaml
from ament_index_python.packages import get_package_share_path
from geometry_msgs.msg import (
    Point,
    Point32,
    Polygon,
    Pose,
    PoseWithCovarianceStamped,
)
from rclpy.executors import MultiThreadedExecutor
from rclpy.node import Node
from rviz_2d_overlay_msgs.msg import OverlayText
from std_msgs.msg import Float32
from std_srvs.srv import Trigger
from tf_transformations import euler_from_quaternion
from visualization_msgs.msg import Marker, MarkerArray

from scenario_msgs.msg import PolygonsStamped, Viewpoint, Viewpoints
from scenario_msgs.srv import MoveToStart


@dataclass
class ViewpointIndex:
    current: int = -1
    previous: int = -1


DEPTH = -0.5


class ScenarioNode(Node):

    def __init__(self):
        super().__init__(node_name='scenario_node')

        self.declare_parameters(
            namespace='',
            parameters=[
                ('scenario', rclpy.Parameter.Type.INTEGER),
                ('viewpoints', rclpy.Parameter.Type.INTEGER),
            ],
        )
        param = self.get_parameter('scenario')
        self.polygons: list[Polygon] = None
        self.scenario = param.value
        self.get_logger().info(f'Using scenario #{self.scenario}.')
        self.read_scenario_description()
        self.read_viewpoints()
        self.init_publishers()
        self.init_clients()
        self.init_services()
        self.init_overlay_texts()
        self.viewpoint_index = ViewpointIndex()
        self.viewpoint_progresses = [
            deque([0.0] * 100, maxlen=100) for _ in self.viewpoints.viewpoints
        ]

        self.pose_sub = self.create_subscription(PoseWithCovarianceStamped,
                                                 'vision_pose_cov',
                                                 self.on_pose, 1)
        self.timer = self.create_timer(1.0 / 50, self.on_timer)
        self.finished_viewpoints = False
        self.running = False
        self.t_start: rclpy.time.Time
        self.start_position_reached = False
        self.vehicle_pose = Pose()

    def init_publishers(self):
        self.obstacles_pub = self.create_publisher(msg_type=PolygonsStamped,
                                                   topic='obstacles',
                                                   qos_profile=1)
        self.viewpoints_pub = self.create_publisher(Viewpoints, 'viewpoints', 1)
        self.gauge_pub = self.create_publisher(Float32, '~/viewpoint_progress',
                                               1)
        self.rviz_overlay_status_pub = self.create_publisher(
            OverlayText, '~/overlay_status_text', 1)
        self.rviz_t_finish_pub = self.create_publisher(
            OverlayText, '~/overlay_t_finish_text', 1)

        self.marker_pub = self.create_publisher(msg_type=MarkerArray,
                                                topic='~/marker_array',
                                                qos_profile=1)

    def init_services(self):
        self.start_service = self.create_service(Trigger, '~/start',
                                                 self.serve_start)
        self.reset_service = self.create_service(Trigger, '~/reset',
                                                 self.serve_reset)

    def init_clients(self):
        # require separate cb group to make synchronous service calls
        cb_group = rclpy.callback_groups.MutuallyExclusiveCallbackGroup()
        self.start_path_planner_client = self.create_client(
            Trigger, 'path_planner/start', callback_group=cb_group)
        self.move_to_start_client = self.create_client(
            MoveToStart, 'path_planner/move_to_start', callback_group=cb_group)
        self.stop_path_planner_client = self.create_client(
            Trigger, 'path_planner/stop', callback_group=cb_group)

    def reset(self):
        for viewpoint in self.viewpoints.viewpoints:
            viewpoint.completed = False
        self.viewpoint_progresses = [
            deque([0.0] * 100, maxlen=100) for _ in self.viewpoints.viewpoints
        ]
        self.start_position_reached = False
        self.running = False
        self.publish_finish_time_overlay(delete=True)

    def serve_start(self, request, response: Trigger.Response):
        if self.running:
            response.success = False
            response.message = "Already running."
            return response
        req = MoveToStart.Request()
        req.current_pose = self.vehicle_pose
        req.target_pose = self.viewpoints.viewpoints[0].pose
        self.move_to_start_client.call(req)
        self.reset()
        self.running = True
        response.success = True
        response.message = "Started."
        return response

    def serve_reset(self, request, response: Trigger.Response):
        self.reset()
        self.stop_path_planner_client.call(Trigger.Request())
        response.success = True
        response.message = "Reset"
        return response

    def init_overlay_texts(self):
        self.init_status_overlay_text()
        self.init_finish_time_overlay_text()

    def init_status_overlay_text(self):
        msg = OverlayText()
        msg.horizontal_alignment = msg.LEFT
        msg.horizontal_distance = 10
        msg.vertical_alignment = msg.TOP
        msg.vertical_distance = 10
        msg.fg_color.a = 1.0
        msg.fg_color.g = 1.0
        msg.line_width = 10
        msg.text_size = 18.0
        msg.width = 1000
        msg.height = 50
        self.status_overlay_text_msg = msg

    def init_finish_time_overlay_text(self):
        msg = OverlayText()
        msg.horizontal_alignment = msg.LEFT
        msg.horizontal_distance = 10
        msg.vertical_alignment = msg.TOP
        msg.vertical_distance = 40
        msg.fg_color.a = 1.0
        msg.fg_color.g = 1.0
        msg.line_width = 10
        msg.text_size = 18.0
        msg.width = 1000
        msg.height = 50
        self.finish_time_overlay_text_msg = msg

    def publish_viewpoint_progress_gauge(self, i):
        if i < 0 or i >= len(self.viewpoint_progresses):
            self.gauge_pub.publish(Float32(data=0.0))
            return
        if self.viewpoints.viewpoints[i].completed:
            self.gauge_pub.publish(Float32(data=0.0))
            return
        if not self.running:
            self.gauge_pub.publish(Float32(data=0.0))
        queue = self.viewpoint_progresses[self.viewpoint_index.previous]
        value = min(1.0, max(0.0, mean(queue) / 0.8))
        self.gauge_pub.publish(Float32(data=value))

    def update_viewpoint_progress(self):
        for j, queue in enumerate(self.viewpoint_progresses):
            if j == self.viewpoint_index.current:
                queue.append(1.0)
            else:
                queue.append(0.0)
        if (not self.start_position_reached) and (self.viewpoint_index.current
                                                  != 0):
            self.viewpoint_progresses[self.viewpoint_index.current][-1] = 0.0

    def on_pose(self, msg: PoseWithCovarianceStamped):
        pose = msg.pose.pose
        self.vehicle_pose = pose
        self.viewpoint_index.current = self.find_viewpoint_in_tolerance(pose)
        self.update_viewpoint_progress()
        if self.viewpoint_index.current < 0:
            self.publish_viewpoint_progress_gauge(self.viewpoint_index.previous)
            return
        self.viewpoint_index.previous = self.viewpoint_index.current
        self.publish_viewpoint_progress_gauge(self.viewpoint_index.current)
        if self.is_viewpoint_completed(
                self.viewpoint_index.current,
                self.viewpoints.viewpoints[self.viewpoint_index.current], pose):
            self.viewpoints.viewpoints[
                self.viewpoint_index.current].completed = True
            self.get_logger().info(f'Viewpoint[{self.viewpoint_index.current}] '
                                   'completed. Uncompleted waypoints: '
                                   f'{self.uncompleted_viewpoints()}')
        uncompleted_waypoints = self.uncompleted_viewpoints()
        if not uncompleted_waypoints:
            self.finished_viewpoints = True
            self.running = False
            t = (self.get_clock().now() - self.t_start).nanoseconds * 1e-9
            self.publish_finish_time_overlay()
            self.get_logger().info(f'Finished after {t:.2f}s')

    def wrap_pi(self, value: float):
        """Normalize the angle to the range [-pi; pi]."""
        if (-math.pi < value) and (value < math.pi):
            return value
        range = 2 * math.pi
        num_wraps = math.floor((value + math.pi) / range)
        return value - range * num_wraps

    def are_all_viewpoints_completed(self):
        x = [True for x in self.viewpoints.viewpoints if not x.completed]
        return not bool(x)

    def uncompleted_viewpoints(self):
        return [
            i for i, x in enumerate(self.viewpoints.viewpoints)
            if not x.completed
        ]

    def find_viewpoint_in_tolerance(self, pose: Pose):
        yaw_tolerance = 0.1
        position_tolerance = 0.1
        p = pose.position
        q = pose.orientation
        _, _, yaw = euler_from_quaternion([q.x, q.y, q.z, q.w])
        for i, viewpoint in enumerate(self.viewpoints.viewpoints):
            # do not consider already completed viewpoints
            if viewpoint.completed:
                continue
            p_target = viewpoint.pose.position
            q = viewpoint.pose.orientation
            _, _, yaw_target = euler_from_quaternion([q.x, q.y, q.z, q.w])
            d = math.sqrt((p.x - p_target.x)**2 + (p.y - p_target.y)**2)
            yaw_error = abs(self.wrap_pi(abs(yaw - yaw_target)))
            if d < position_tolerance and yaw_error < yaw_tolerance:
                return i
        return -1

    def is_viewpoint_completed(self, i, viewpoint: Viewpoint, pose: Pose):
        # ignore waypoints until start position has been reached
        if not self.start_position_reached:
            if i != 0:
                self.publish_viewpoint_progress_gauge(-1)
                return
        completed = mean(self.viewpoint_progresses[i]) >= 0.8
        if completed and not self.start_position_reached:
            self.start_position_reached = True
            self.t_start = self.get_clock().now()
            self.start_path_planner_client.call(Trigger.Request())
        return completed

    def publish_marker_array(self, markers):
        self.marker_pub.publish(MarkerArray(markers=markers))

    def create_polygon_markers(self, polygons):
        stamp = self.get_clock().now().to_msg()
        markers = []
        for i, polygon in enumerate(polygons):
            marker = Marker()
            for p in polygon.points:
                marker.points.append(Point(x=p.x, y=p.y, z=p.z))
            marker.points.append(marker.points[0])
            marker.type = Marker.LINE_STRIP
            if not self.running:
                marker.action = Marker.DELETEALL
            else:
                marker.action = Marker.ADD
            marker.id = i
            marker.color.r = 1.0
            marker.color.a = 1.0
            marker.scale.x = 0.02
            marker.scale.y = 0.02
            marker.scale.z = 0.02
            marker.header.stamp = stamp
            marker.header.frame_id = 'map'
            marker.ns = 'obstacles'
            markers.append(marker)
        return markers

    def create_viewpoint_marker(self):
        markers = []
        viewpoint: Viewpoint
        stamp = self.get_clock().now().to_msg()
        for i, viewpoint in enumerate(self.viewpoints.viewpoints):
            marker = Marker()
            marker.header.stamp = stamp
            marker.header.frame_id = 'map'
            marker.pose = viewpoint.pose
            if not self.running:
                marker.action = Marker.DELETEALL
            else:
                marker.action = Marker.ADD
            marker.type = Marker.ARROW
            marker.id = i
            marker.ns = 'viewpoints'
            marker.color.a = 1.0
            marker.color.r = 1.0 - viewpoint.completed
            marker.color.g = 1.0 * viewpoint.completed
            marker.scale.x = 0.3
            marker.scale.y = 0.1
            marker.scale.z = 0.1
            markers.append(marker)
        i = self.viewpoint_index.current
        if i >= 0 and not self.viewpoints.viewpoints[i].completed:
            markers[i].color.r = 1.0
            markers[i].color.g = 1.0
            markers[i].color.b = 0.0

        return markers

    def publish_finish_time_overlay(self, delete=False):
        msg = self.finish_time_overlay_text_msg
        msg.action = OverlayText.ADD
        if delete:
            msg.action = OverlayText.DELETE
        else:
            t = (self.get_clock().now() - self.t_start).nanoseconds * 1e-9
            msg.text = f'Finished after {t:.2f}s'
        self.rviz_t_finish_pub.publish(msg)

    def publish_status_overlay(self):
        msg = self.status_overlay_text_msg
        if self.running:
            msg.fg_color.r = 1.0
            msg.fg_color.g = 1.0
            msg.fg_color.b = 0.0
            if self.start_position_reached:
                t = (self.get_clock().now() - self.t_start).nanoseconds * 1e-9
                msg.text = f'Time: {t:.2f}s'
            else:
                msg.text = 'Moving to start position'
        else:
            msg.fg_color.r = 1.0
            msg.fg_color.g = 0.0
            msg.fg_color.b = 0.0
            msg.text = 'Not running'
        self.rviz_overlay_status_pub.publish(msg)

    def on_timer(self):
        polygons_msg = PolygonsStamped()
        polygons_msg.header.stamp = self.get_clock().now().to_msg()
        polygons_msg.header.frame_id = 'map'
        polygons_msg.polygons = self.polygons

        if self.running:
            self.obstacles_pub.publish(polygons_msg)
            self.viewpoints_pub.publish(self.viewpoints)

        self.publish_status_overlay()
        markers = self.create_polygon_markers(self.polygons)
        markers.extend(self.create_viewpoint_marker())
        self.publish_marker_array(markers)

    def read_scenario_description(self):
        filepath = os.path.join(get_package_share_path('final_project'),
                                f'config/scenario_{self.scenario}.yaml')
        self.get_logger().info(filepath)

        def obstacle_to_polygon(obstacle):
            polygon = Polygon()
            for corner in obstacle['corners']:
                p = Point32(x=corner[0], y=corner[1], z=DEPTH)
                polygon.points.append(p)
            return polygon

        polygons = []
        with open(filepath, 'r') as f:
            data = yaml.safe_load(f)
            for obstacle in data['obstacles']:
                polygon = obstacle_to_polygon(obstacle)
                polygons.append(polygon)
        self.polygons = polygons

    def read_viewpoints(self):
        path = os.path.join(get_package_share_path('final_project'),
                            f'config/scenario_{self.scenario}.yaml')
        with open(path, 'r') as f:
            data = yaml.safe_load(f)
            self.viewpoints = Viewpoints()
            for waypoint in data['viewpoints']:
                pose = Pose()
                pose.position.x = waypoint['x']
                pose.position.y = waypoint['y']
                pose.position.z = waypoint['z']
                pose.orientation.w = waypoint['qw']
                pose.orientation.x = waypoint['qx']
                pose.orientation.y = waypoint['qy']
                pose.orientation.z = waypoint['qz']
                viewpoint = Viewpoint()
                viewpoint.pose = pose
                viewpoint.completed = False
                self.viewpoints.viewpoints.append(viewpoint)


def main():
    rclpy.init()
    node = ScenarioNode()
    exec = MultiThreadedExecutor()
    exec.add_node(node)
    try:
        exec.spin()
    except KeyboardInterrupt:
        pass
    node.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: fav/src/scenario/scenario_node.cpp
#include "scenario_node.hpp"

#include <hippo_common/convert.hpp>
#include <hippo_common/param_utils.hpp>
#include <rclcpp/experimental/executors/events_executor/events_executor.hpp>
namespace fav {
namespace scenario {

static constexpr double kDepth = -0.5;

ScenarioNode::ScenarioNode() : Node("scenario_node") {
  HIPPO_COMMON_DECLARE_PARAM_READONLY(scenario);
  if (!ReadConfig(params_.scenario)) {
    RCLCPP_FATAL(get_logger(), "Failed to load config.");
    return;
  }
  PopulateViewpointsMessage();

  InitPublishers();
  InitClients();
  InitSubscriptions();
  InitServices();
  InitTimers();
}

void ScenarioNode::InitPublishers() {
  std::string topic;
  rclcpp::QoS qos = rclcpp::SystemDefaultsQoS();
  qos.keep_last(1);

  topic = "obstacles";
  obstacles_pub_ =
      create_publisher<scenario_msgs::msg::PolygonsStamped>(topic, qos);

  topic = "viewpoints";
  viewpoints_pub_ =
      create_publisher<scenario_msgs::msg::Viewpoints>(topic, qos);

  topic = "~/viewpoint_progress";
  progress_pub_ = create_publisher<std_msgs::msg::Float32>(topic, qos);

  topic = "~/overlay_t_finish_text";
  finish_time_overlay_pub_ =
      create_publisher<rviz_2d_overlay_msgs::msg::OverlayText>(topic, qos);

  topic = "~/overlay_status_text";
  status_overlay_pub_ =
      create_publisher<rviz_2d_overlay_msgs::msg::OverlayText>(topic, qos);

  topic = "~/marker_array";
  marker_array_pub_ =
      create_publisher<visualization_msgs::msg::MarkerArray>(topic, qos);
}

void ScenarioNode::InitSubscriptions() {
  std::string topic;
  rclcpp::QoS qos = rclcpp::SensorDataQoS();
  qos.keep_last(1);

  topic = "vision_pose_cov";
  using geometry_msgs::msg::PoseWithCovarianceStamped;
  pose_sub_ = create_subscription<PoseWithCovarianceStamped>(
      topic, qos,
      [this](const PoseWithCovarianceStamped::SharedPtr msg) { OnPose(msg); });
}

void ScenarioNode::InitTimers() {
  timer_ = create_timer(std::chrono::milliseconds(20), [this]() { OnTimer(); });
}

void ScenarioNode::InitClients() {
  std::string name;
  client_cb_group_ = create_callback_group(
      rclcpp::CallbackGroupType::MutuallyExclusive, false);
  rclcpp::QoS qos = rclcpp::ServicesQoS();

  name = "path_planner/start";
  start_path_planner_client_ =
      create_client<std_srvs::srv::Trigger>(name, qos, client_cb_group_);

  name = "path_planner/move_to_start";
  move_to_start_client_ = create_client<scenario_msgs::srv::MoveToStart>(
      name, qos, client_cb_group_);

  name = "path_planner/stop";
  stop_path_planner_client_ =
      create_client<std_srvs::srv::Trigger>(name, qos, client_cb_group_);
}

void ScenarioNode::InitServices() {
  std::string name;
  using std_srvs::srv::Trigger;

  name = "~/start";
  start_service_ = create_service<Trigger>(
      name, [this](const Trigger::Request::SharedPtr request,
                   Trigger::Response::SharedPtr response) {
        ServeStart(request, response);
      });

  name = "~/reset";
  reset_service_ = create_service<Trigger>(
      name, [this](const Trigger::Request::SharedPtr request,
                   Trigger::Response::SharedPtr response) {
        ServeReset(request, response);
      });
}

void ScenarioNode::ServeStart(
    const std_srvs::srv::Trigger::Request::SharedPtr,
    std_srvs::srv::Trigger::Response::SharedPtr response) {
  if (running_) {
    response->success = false;
    response->message = "Already running.";
    return;
  }
  Reset();
  running_ = true;
  using hippo_common::convert::EigenToRos;
  auto req = std::make_shared<scenario_msgs::srv::MoveToStart_Request>();
  EigenToRos(vehicle_position_, req->current_pose.position);
  EigenToRos(vehicle_orientation_, req->current_pose.orientation);
  req->target_pose = viewpoints_msg_.viewpoints.at(0).pose;
  if (!move_to_start_client_) {
    RCLCPP_ERROR(get_logger(), "move_to_start service client not initialized");
    return;
  }
  auto future = move_to_start_client_->async_send_request(req);
  auto status = future.wait_for(std::chrono::seconds(1));
  if (status == std::future_status::timeout) {
    // we have to remove pending service calls manually according to the docs.
    move_to_start_client_->remove_pending_request(future);
    RCLCPP_WARN(get_logger(),
                "Failed to make a call to the path_planner/move_to_start "
                "service. The scenario will start nonetheless. But the "
                "path_planner will probably not work.");
  }
  response->success = true;
  response->message = "Started";
}
void ScenarioNode::ServeReset(
    const std_srvs::srv::Trigger::Request::SharedPtr,
    std_srvs::srv::Trigger::Response::SharedPtr response) {
  Reset();
  auto req = std::make_shared<std_srvs::srv::Trigger_Request>();
  if (!stop_path_planner_client_) {
    RCLCPP_ERROR(get_logger(),
                 "Service for starting the path planner not initialied.");
    return;
  }
  auto future = stop_path_planner_client_->async_send_request(req);
  auto status = future.wait_for(std::chrono::seconds(1));
  if (status == std::future_status::timeout) {
    stop_path_planner_client_->remove_pending_request(future);
    RCLCPP_WARN(get_logger(),
                "Failed to make a call to the path_planner/stop service. The "
                "scenario will start nonetheless. But the path_planner will "
                "probably not work.:");
  }
  response->success = true;
  response->message = "Reset";
}

void ScenarioNode::PublishProgressGauge() {
  int i = viewpoint_index_.current < 0 ? viewpoint_index_.previous
                                       : viewpoint_index_.current;
  std_msgs::msg::Float32 msg;
  if (i >= 0 && !viewpoints_.at(i).IsCompleted()) {
    msg.data = viewpoints_.at(i).Progress();
  }
  progress_pub_->publish(msg);
}

void ScenarioNode::PublishFinishTimeOverlay(double _dt, bool delete_text) {
  rviz_2d_overlay_msgs::msg::OverlayText msg;
  msg.action = delete_text ? msg.DELETE : msg.ADD;
  msg.horizontal_alignment = msg.LEFT;
  msg.horizontal_distance = 10;
  msg.vertical_alignment = msg.TOP;
  msg.vertical_distance = 40;
  msg.fg_color.a = 1.0;
  msg.fg_color.g = 1.0;
  msg.line_width = 10;
  msg.text_size = 18.0;
  msg.width = 1000;
  msg.height = 50;

  char buffer[256];
  snprintf(buffer, sizeof(buffer), "Finished after %.2lf", _dt);
  msg.text = buffer;
  finish_time_overlay_pub_->publish(msg);
}

void ScenarioNode::PublishStatusOverlay() {
  rviz_2d_overlay_msgs::msg::OverlayText msg;
  msg.horizontal_alignment = msg.LEFT;
  msg.horizontal_distance = 10;
  msg.vertical_alignment = msg.TOP;
  msg.vertical_distance = 10;
  msg.fg_color.a = 1.0;
  msg.fg_color.g = 1.0;
  msg.line_width = 10;
  msg.text_size = 18.0;
  msg.width = 1000;
  msg.height = 50;
  if (running_) {
    msg.fg_color.r = 1.0;
    msg.fg_color.g = 1.0;
    msg.fg_color.b = 0.0;
    if (start_position_reached_) {
      double dt = (now() - t_start_).nanoseconds() * 1e-9;
      char buffer[256];
      snprintf(buffer, sizeof(buffer), "Time: %.2lf", dt);
      msg.text = buffer;
    } else {
      msg.text = "Moving to start position";
    }
  } else {
    msg.fg_color.r = 1.0;
    msg.fg_color.g = 0.0;
    msg.fg_color.b = 0.0;
    msg.text = "Not running";
  }
  status_overlay_pub_->publish(msg);
}

void ScenarioNode::PublishMarkerArray() {
  marker_array_.markers.clear();
  AddObstacleMarkers();
  AddViewpointMarkers();
  if (!marker_array_pub_) {
    RCLCPP_ERROR(get_logger(), "Marker array publisher not initialized.");
    return;
  }
  marker_array_pub_->publish(marker_array_);
}

void ScenarioNode::AddObstacleMarkers() {
  visualization_msgs::msg::Marker marker;
  int i = 0;
  marker.header.stamp = now();
  marker.header.frame_id = "map";
  for (const auto &obstacle : obstacles_) {
    marker.points.clear();
    for (const auto &point : obstacle.points) {
      geometry_msgs::msg::Point p;
      p.x = point.x;
      p.y = point.y;
      p.z = point.z;
      marker.points.push_back(p);
    }
    marker.type = marker.LINE_STRIP;
    marker.action = marker.ADD;
    if (!running_) {
      marker.action = marker.DELETEALL;
    }
    marker.id = i++;
    marker.ns = "obstacles";
    marker.color.r = 1.0;
    marker.color.a = 1.0;
    marker.scale.x = 0.02;
    marker.scale.y = 0.02;
    marker.scale.z = 0.02;
    marker_array_.markers.push_back(marker);
  }
}

void ScenarioNode::AddViewpointMarkers() {
  visualization_msgs::msg::Marker marker;
  marker.header.stamp = now();
  marker.header.frame_id = "map";
  marker.type = marker.ARROW;
  marker.ns = "viewpoints";
  marker.scale.x = 0.3;
  marker.scale.y = 0.1;
  marker.scale.z = 0.1;
  marker.color.a = 1.0;
  int i = 0;
  const auto t_now = now();
  for (const auto &viewpoint : viewpoints_msg_.viewpoints) {
    marker.pose = viewpoint.pose;
    marker.action = marker.ADD;
    marker.id = i++;
    if (!running_) {
      marker.action = marker.DELETEALL;
    }
    marker.color.r = 1.0 - viewpoint.completed;
    marker.color.g = 1.0 * viewpoint.completed;
    marker_array_.markers.push_back(marker);
    if (i == viewpoint_index_.current && !viewpoint.completed) {
      marker.color.r = 1.0;
      marker.color.g = 1.0;
      marker.color.b = 0.0;
    }
  }
}

void ScenarioNode::OnPose(
    const geometry_msgs::msg::PoseWithCovarianceStamped::SharedPtr _msg) {
  using hippo_common::convert::RosToEigen;
  RosToEigen(_msg->pose.pose.position, vehicle_position_);
  RosToEigen(_msg->pose.pose.orientation, vehicle_orientation_);

  UpdateViewpointProgress();
  viewpoint_index_.current = FindViewpointInTolerance();
  PopulateViewpointsMessage();
  if (viewpoint_index_.current >= 0) {
    viewpoint_index_.previous = viewpoint_index_.current;
  }

  PublishProgressGauge();
  if (running_ && viewpoints_finished_) {
    running_ = false;
    double dt = (now() - t_start_).nanoseconds() * 1e-9;
    PublishFinishTimeOverlay(dt);
    RCLCPP_INFO(get_logger(), "Finished after %.2lfs", dt);
  }
}

void ScenarioNode::UpdateViewpointProgress() {
  if (!viewpoints_.at(0).IsCompleted()) {
    // still moving to start location. no progress of other viewpoints allowed
    viewpoints_[0].Update(vehicle_position_, vehicle_orientation_);
    if (viewpoints_[0].IsCompleted()) {
      start_position_reached_ = true;
      t_start_ = now();
      auto req = std::make_shared<std_srvs::srv::Trigger_Request>();
      auto future = start_path_planner_client_->async_send_request(req);
      auto status = future.wait_for(std::chrono::milliseconds(500));
      if (status == std::future_status::timeout) {
        RCLCPP_WARN(
            get_logger(),
            "Could not call path_planner/start service. Scenario will continue "
            "nonetheless. But probably things aren't working.");
      }
    }
    return;
  }
  bool progress_disabled{false};
  viewpoints_finished_ = true;
  for (auto &viewpoint : viewpoints_) {
    if (viewpoint.IsCompleted()) {
      continue;
    }
    viewpoints_finished_ = false;
    if (progress_disabled) {
      viewpoint.Update(false);
    } else {
      viewpoint.Update(vehicle_position_, vehicle_orientation_);
      if (viewpoint.IsInTolerance()) {
        // only let the first viewpoint in tolerance continue its progress.
        progress_disabled = true;
      }
    }
  }
}

int ScenarioNode::FindViewpointInTolerance() {
  for (size_t i = 0; i < viewpoints_.size(); ++i) {
    if (viewpoints_.at(i).IsCompleted()) {
      continue;
    }
    if (viewpoints_.at(i).IsInTolerance()) {
      return i;
    }
  }
  return -1;
}

void ScenarioNode::OnTimer() {
  scenario_msgs::msg::PolygonsStamped obstacles_msg;
  obstacles_msg.header.stamp = now();
  obstacles_msg.header.frame_id = "map";
  obstacles_msg.polygons = obstacles_;

  if (running_) {
    if (obstacles_pub_) {
      obstacles_pub_->publish(obstacles_msg);
    }
    if (viewpoints_pub_) {
      viewpoints_msg_.header.stamp = now();
      viewpoints_msg_.header.frame_id = "map";
      viewpoints_pub_->publish(viewpoints_msg_);
    }
  }
  PublishStatusOverlay();
  PublishMarkerArray();
}

void ScenarioNode::Reset() {
  for (auto &viewpoint : viewpoints_) {
    viewpoint.Reset();
  }
  PopulateViewpointsMessage();
  running_ = false;
  start_position_reached_ = false;
  viewpoints_finished_ = false;
  // TODO delete the finish time overlay
}

void ScenarioNode::PopulateViewpointsMessage() {
  using hippo_common::convert::EigenToRos;
  viewpoints_msg_.viewpoints.clear();
  for (auto const &viewpoint : viewpoints_) {
    scenario_msgs::msg::Viewpoint viewpoint_msg;
    geometry_msgs::msg::Pose p;
    EigenToRos(viewpoint.position, p.position);
    EigenToRos(viewpoint.orientation, p.orientation);

    viewpoint_msg.pose = p;
    viewpoint_msg.progress = viewpoint.Progress();
    viewpoint_msg.completed = viewpoint.IsCompleted();
    viewpoints_msg_.viewpoints.push_back(viewpoint_msg);
  }
}

std::optional<std::string> ScenarioNode::GetScenarioFilePath(int scenario) {
  std::string file_path;
  std::string pkg{"final_project"};
  try {
    file_path = ament_index_cpp::get_package_share_directory(pkg);
  } catch (const ament_index_cpp::PackageNotFoundError &) {
    RCLCPP_FATAL(
        get_logger(),
        "Failed to load scenario config because [%s] could not be found.",
        pkg.c_str());
    return std::nullopt;
  }
  file_path += "/config/scenario_" + std::to_string(scenario) + ".yaml";
  return file_path;
}

bool ScenarioNode::ReadConfig(int scenario) {
  std::optional<std::string> file_path = GetScenarioFilePath(scenario);
  if (!file_path) {
    return false;
  }
  YAML::Node node;
  try {
    node = YAML::LoadFile(*file_path);
  } catch (const YAML::BadFile &) {
    RCLCPP_FATAL(
        get_logger(),
        "Could not load scenario config because file [%s] could not be "
        "loaded. Does it exist?",
        file_path->c_str());
    return false;
  } catch (const YAML::ParserException &) {
    RCLCPP_FATAL(
        get_logger(),
        "Could not load scenario config from [%s] because it is malformed.",
        file_path->c_str());
    return false;
  }
  ReadViewpoints(node);
  ReadObstacles(node);
  return true;
}

bool ScenarioNode::ReadObstacles(const YAML::Node &node) {
  using geometry_msgs::msg::Point32;
  using geometry_msgs::msg::Polygon;
  obstacles_.clear();
  for (const YAML::Node &obstacle : node["obstacles"]) {
    std::vector<Point32> corners;
    Polygon polygon;
    for (const YAML::Node &corner : obstacle["corners"]) {
      std::vector<double> corner_points = corner.as<std::vector<double>>();
      Point32 p;
      p.x = corner_points.at(0);
      p.y = corner_points.at(1);
      p.z = kDepth;
      polygon.points.push_back(p);
    }
    polygon.points.push_back(polygon.points.at(0));
    obstacles_.push_back(polygon);
  }
  return true;
}

bool ScenarioNode::ReadViewpoints(const YAML::Node &node) {
  viewpoint_poses_ = node["viewpoints"].as<hippo_common::yaml::Poses>();
  viewpoints_.clear();
  for (const auto &pose : viewpoint_poses_) {
    viewpoints_.emplace_back(pose.position, pose.orientation, 0.1, 0.2);
  }
  return true;
}
}  // namespace scenario
}  // namespace fav

int main(int argc, char **argv) {
  rclcpp::init(argc, argv);
  std::vector<std::shared_ptr<std::thread>> spin_threads;
  rclcpp::experimental::executors::EventsExecutor exec_main;
  rclcpp::experimental::executors::EventsExecutor exec_client;
  auto node = std::make_shared<fav::scenario::ScenarioNode>();
  exec_main.add_node(node->get_node_base_interface());
  exec_client.add_callback_group(node->client_cb_group_,
                                 node->get_node_base_interface());
  spin_threads.push_back(
      std::make_shared<std::thread>([&exec_main]() { exec_main.spin(); }));
  spin_threads.push_back(
      std::make_shared<std::thread>([&exec_client]() { exec_client.spin(); }));

  for (auto &thread : spin_threads) {
    thread->join();
  }
  rclcpp::shutdown();
  return 0;
}
